ts_src/psetv2/finalizer.ts

```diff
@@ -0,0 +1,272 @@
+    if (
+      (!input.partialSigs || input.partialSigs!.length === 0) &&
+      (!input.tapKeySig || input.tapKeySig.length === 0) &&
+      (!input.tapScriptSig || input.tapScriptSig.length === 0)
+    ) {
+      throw new Error('Missing input partial signatures');
+    }
+
+    const pset = this.pset.copy();
+
+    const { finalScriptSig, finalScriptWitness } = finalizeFunc(inIndex, pset);
+
+    if (finalScriptSig) {
+      pset.inputs[inIndex].finalScriptSig = finalScriptSig;
+    }
+    if (finalScriptWitness) {
+      pset.inputs[inIndex].finalScriptWitness = finalScriptWitness;
+    }
+    if (!finalScriptSig && !finalScriptWitness && !input.finalScriptWitness) {
+      throw new Error(`Unknown error finalizing input #${inIndex}`);
+    }
+
+    pset.sanityCheck();
+
+    this.pset.globals = pset.globals;
+    this.pset.inputs = pset.inputs;
+    this.pset.outputs = pset.outputs;
+
+    return this;
+  }
+}
+
+interface GetScriptReturn {
+  script: Buffer | null;
+  isSegwit: boolean;
+  isP2SH: boolean;
+  isP2WSH: boolean;
+}
+
+function getScriptFromInput(input: Input): GetScriptReturn {
+  const res: GetScriptReturn = {
+    script: null,
+    isSegwit: false,
+    isP2SH: false,
+    isP2WSH: false,
+  };
+  res.isP2SH = !!input.redeemScript;
+  res.isP2WSH = !!input.witnessScript;
+  if (input.witnessScript) {
+    res.script = input.witnessScript!;
+  } else if (input.redeemScript) {
+    res.script = input.redeemScript!;
+  } else {
+    if (input.nonWitnessUtxo) {
+      res.script = input.nonWitnessUtxo!.outs[input.previousTxIndex].script;
+    } else if (input.witnessUtxo) {
+      res.script = input.witnessUtxo!.script;
+    }
+  }
+  if (input.witnessScript || isP2WPKH(res.script!)) {
+    res.isSegwit = true;
+  }
+  return res;
+}
+
+const defaultFinalizer: FinalizeFunc = (inIndex: number, pset: Pset) => {
+  const input = pset.inputs[inIndex];
+  if (input.isTaproot()) return finalizeTaprootInput(inIndex, pset);
+  return finalizeInput(inIndex, pset);
+};
+
+const finalizeInput = (inIndex: number, pset: Pset) => {
+  const input = pset.inputs[inIndex];
+  const { script, isP2SH, isP2WSH, isSegwit } = getScriptFromInput(input);
+  if (!script) throw new Error(`No script found for input #${inIndex}`);
+
+  return getFinalScripts(inIndex, input, script, isSegwit, isP2SH, isP2WSH);
+};
+
+const finalizeTaprootInput = (
+  inIndex: number,
+  pset: Pset,
+): {
+  finalScriptSig?: Buffer;
+  finalScriptWitness?: Buffer;
+} => {
+  const input = pset.inputs[inIndex];
+  if (!input.witnessUtxo)
+    throw new Error(
+      `Cannot finalize input #${inIndex}. Missing withness utxo.`,
+    );
+
+  // Check key spend first. Increased privacy and reduced block space.
+  if (input.tapKeySig) {
+    return {
+      finalScriptWitness: witnessStackToScriptWitness([input.tapKeySig!]),
+    };
+  } else {
+    return getTaprootFinalScripts(
+      inIndex,
+      input,
+      // tapLeafHashToFinalize, // TODO: what about this?
+    );
```

Comment:
need help on this

Reply:
We need a way to select the tapLeaf to finalize in case of several `input.tapLeafScript` are set up. However, `finalizeTaprootInput` seems to be used only as a default finalizer -> by default, I would always finalize the first `input.tapLeafScript[0]` because not sure several leaves scenarios are really common (most of the time u don't want to share several leaves with others :thinking: ?). 

If the user wants to handle several leaves scenario he needs to write another finalizer func (with a tapLeafTofinalize argument for instance)

Reply:
allright, so that field is for "non-default" cases, which, as you correctly say, require a custom finalization func. I'm gonna remove the commented line then. Thanks!